### PR TITLE
AuthZ Service: Add caching

### PR DESCRIPTION
**Pull request by gamab**
This pull request introduces caching mechanisms to improve the performance of user permission checks in the RBAC service. The most important changes include adding cache services for user permissions, basic roles, and team memberships, refactoring the `Store` interface, and updating the test suite to cover these new functionalities.

### CODE REVIEW

File: pkg/services/authz/rbac/service_test.go [hunk 1]
```diff
@@ -143,9 +149,294 @@ func TestService_checkPermission(t *testing.T) {
...
+			if tc.cacheHit {
+				require.Equal(t, 1, store.calls) // only get user id
+			} else {
+				require.Equal(t, 3, store.calls) // get user id, basic role, and permissions
+			}
+		})
+	}
+}
+
+type fakeStore struct {
```
Comment (IevaVasiljeva) [inline pkg/services/authz/rbac/service_test.go:389]: nit: I've got a slight preference for having the fake store in a separate file to keep the test file a bit cleaner and allow for the fake store to be reused (unlikely in this case tbh). But also fine to keep it as it is if you like it better this way.

**Review by IevaVasiljeva — APPROVED** [review]
Nice, looks good to me! I think it might also be worth caching user UID -> ID mapping, but apart from that all looks good 👌

File: pkg/services/authz/rbac/service.go [hunk 2]
```diff
@@ -22,25 +24,40 @@ import (
...
 
-	store         *store.Store
+	store         store.Store
 	identityStore legacy.LegacyIdentityStore
 	actionMapper  *mappers.K8sRbacMapper
 
 	logger log.Logger
 	tracer tracing.Tracer
+
+	// Cache for user permissions, user team memberships and user basic roles
```
[the remark this answers is not in the document]
Comment (gamab) [reply pkg/services/authz/rbac/service.go:45]: Let's do it.

File: pkg/services/authz/rbac/service_test.go (hunk 1 above)
Comment (gamab) [reply pkg/services/authz/rbac/service_test.go:389]: Fine by me :)

**Comment by gamab** [comment]
Last commit doesn't make sense actually as we will resolve the identity each time we have a different action. So will revert and change

File: pkg/services/authz/rbac/service.go [hunk 3]
```diff
@@ -140,12 +207,19 @@ func (s *Service) getUserPermissions(ctx context.Context, req *CheckRequest) ([]
 		return nil, fmt.Errorf("could not get user internal id: %w", err)
 	}
 
-	basicRoles, err := s.store.GetBasicRoles(ctx, req.Namespace, store.BasicRoleQuery{UserID: userIdentifiers.ID})
-	if err != nil {
-		return nil, fmt.Errorf("could not get basic roles: %w", err)
-	}
+	s.idCache.Set(uidCacheKey, userIdentifiers, 0)
```
Comment (gamab) [inline pkg/services/authz/rbac/service.go:210]: Not really best practice, but as we are storing twice, I'm storing a pointer. Wdyt?

**Pull request merged**